[DOC] Character filters - Pattern replace (#8557)
opensearch-project/documentation-website: _analyzers/character-filters/pattern-replace-character-filter.md

- `[ ]`: Defines a **character class**, meaning it will match **any one** of the characters inside the brackets.
- `\\s`: Matches any **white space** character, such as a space, tab, or newline.
- `()`: Matches literal **parentheses** (`(` or `)`).
- `-`: Matches a literal **hyphen** (`-`).
- `+`: Specifies that the pattern should match **one or more** occurrences of the preceding characters.

The pattern `[\\s()-]+` will match any sequence of one or more white space characters, parentheses, or hyphens, removing from the input text. This ensures the phone numbers are normalized and contain only digits.

The following request standardizes phone numbers by removing spaces, dashes, and parentheses: 

```json
GET /_analyze
{
  "tokenizer": "standard",
  "char_filter": [
    {
      "type": "pattern_replace",
      "pattern": "[\\s()-]+",
      "replacement": ""
    }
  ],
  "text": "(555) 123-4567"
}
```
{% include copy-curl.html %}

The response contains the generated token:

```json
{
  "tokens": [
    {
      "token": "5551234567",
      "start_offset": 1,
      "end_offset": 14,
      "type": "<NUM>",
      "position": 0
    }
  ]
}
```
 
## Parameters

> The html_strip character filter must be configured with the following parameters.

Reviewer: ```suggestion
The `pattern_replace` character filter must be configured with the following parameters.
```

| Parameter   | Required/Optional | Data type | Description    |
|:---|:---|

> | `pattern`   | Required | String | A regular expression used to match parts of the input text. The filter identifies and matches this pattern for replacement. |

Reviewer: ```suggestion
| `pattern`   | Required | String | A regular expression used to match parts of the input text. The filter identifies and matches this pattern to perform replacement. |
```

          "tokenizer": "standard",
          "char_filter": [
            "pattern_char_filter"
          ]
        }
      },
      "char_filter": {
        "pattern_char_filter": {
          "type": "pattern_replace",
          "pattern": "[$€,.]",
          "replacement": ""
        }
      }
    }
  }
}
```

{% include copy-curl.html %}

Use the following request to examine the tokens generated using the analyzer:

```json
POST /my_index/_analyze
{
  "analyzer": "my_analyzer",
  "text": "Total: $ 1,200.50 and € 1.100,75"
}
```
{% include copy-curl.html %}

The response contains the generated tokens:

```json
{
  "tokens": [
    {
      "token": "Total",
      "start_offset": 0,
      "end_offset": 5,
      "type": "<ALPHANUM>",
      "position": 0
    },
    {
      "token": "120050",
      "start_offset": 9,
      "end_offset": 17,
      "type": "<NUM>",
      "position": 1
    },
    {
      "token": "and",
      "start_offset": 18,
      "end_offset": 21,
      "type": "<ALPHANUM>",
      "position": 2
    },
    {
      "token": "110075",
      "start_offset": 24,
      "end_offset": 32,
      "type": "<NUM>",
      "position": 3
    }
  ]
}
```

## Using capturing groups

> You can use capturing groups in `replacement`. For example, the following request creates a custom analyzer that uses a `pattern_replace` character filter to replace hyphens with dots in phone numbers:

Reviewer: "in `replacement`" => "with the `replacement` parameter"?

Reviewer 2: "in" is better. Reworded.

> You can use capturing groups in `replacement`. For example, the following request creates a custom analyzer that uses a `pattern_replace` character filter to replace hyphens with dots in phone numbers:

Reviewer 2: ```suggestion
You can use capturing groups in the `replacement` parameter. For example, the following request creates a custom analyzer that uses a `pattern_replace` character filter to replace hyphens with dots in phone numbers:
```

title: Pattern Replace Character Filter
parent: Character Filters
nav_order: 130
---

# Pattern Replace Character Filter

The `pattern replace` character filter allows you to use regular expressions (regex) to define patterns for matching and replacing characters in the input text. It is a flexible tool for advanced text transformations, especially when dealing with complex string patterns.

This filter replaces all instances of a pattern with a specified replacement string, allowing for easy substitutions, deletions, or complex modifications to the input text. You can use it to normalize the input before tokenization.

## Example of the pattern replacement character filter

The following example standardizes phone numbers by removing spaces, dashes, or parentheses. 

```json
GET /_analyze
{
  "tokenizer": "standard",
  "char_filter": [
    {
      "type": "pattern_replace",
      "pattern": "[\\s()-]+",
      "replacement": ""
    }
  ],
  "text": "Call me at (555) 123-4567 or 555 987 6543"
}
```

Running this `pattern replace` filter to remove spaces, dashes, and parentheses by replacing the pattern `[\\s()-]+` with an empty string will result in the following text:

```

> Call me at 5551234567 or 5559876543 

Reviewer 3: This won't be the out put as we are relacing all the whitespaces with ""

Correct output would be concatenated string

```
Callmeat5551234567or5559876543
```
